src/main/java/io/confluent/connect/jdbc/source/TimestampTableQuerier.java

```diff
@@ -0,0 +1,205 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.connect.jdbc.source;
+
+import java.util.TimeZone;
+
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.util.List;
+import java.util.Map;
+
+import io.confluent.connect.jdbc.dialect.DatabaseDialect;
+import io.confluent.connect.jdbc.source.SchemaMapping.FieldSetter;
+
+/**
+ * A specialized subclass of the {@link TimestampIncrementingTableQuerier} that only advances the
+ * to-be-committed timestamp offset for its records after either all rows have been read from the
+ * current table query, or all rows with the to-be-committed timestamp have been read successfully
+ * (and the next row has a timestamp that is strictly greater than it).
+ * This prevents data loss in cases where the table has multiple rows with identical timestamps and
+ * the connector is shut down in the middle of reading these rows. However, if the connector is
+ * configured with additional query logic (such as a suffix containing a LIMIT clause), data loss
+ * may still occur, since the timestamp offset for the last row of each query is unconditionally
+ * committed.
+ */
+public class TimestampTableQuerier extends TimestampIncrementingTableQuerier {
+  private static final Logger log = LoggerFactory.getLogger(TimestampTableQuerier.class);
+
+  private boolean exhaustedResultSet;
+  private PendingRecord nextRecord;
+  private Timestamp latestCommittableTimestamp;
+
+  public TimestampTableQuerier(
+      DatabaseDialect dialect,
+      QueryMode mode,
+      String name,
+      String topicPrefix,
+      List<String> timestampColumnNames,
+      Map<String, Object> offsetMap,
+      Long timestampDelay,
+      TimeZone timeZone
+  ) {
+    super(
+        dialect,
+        mode,
+        name,
+        topicPrefix,
+        timestampColumnNames,
+        null,
+        offsetMap,
+        timestampDelay,
+        timeZone
+    );
+
+    this.latestCommittableTimestamp = this.offset.getTimestampOffset();
+    this.exhaustedResultSet = false;
+    this.nextRecord = null;
+  }
+
+  @Override
+  public boolean next() throws SQLException {
+    if (exhaustedResultSet && nextRecord == null) {
```

Comment:
This is to avoid invoking `ResultSet::next` more than once on an exhausted result set. According to the [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#next--):

> When a call to the next method returns `false`, the cursor is positioned after the last row.... If the result set type is `TYPE_FORWARD_ONLY`, it is vendor specified whether their JDBC driver implementation will return `false` or throw an `SQLException` on a subsequent call to `next`.

The default result set type is `TYPE_FORWARD_ONLY` (see [Connection::createStatement](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#createStatement--)), and the connector doesn't manually override the type, so we are using that result set type at the moment.

Reply:
While I understand that consideration, do we really care to provide a more robust contract in our own `next()` method than what the JDBC driver implements? For all practical purposes, we should not be calling this method multiple times on an exhausted result set anyways, right?

Reply:
That's all fair; I'm just not sure there's a way to distinguish between an exhausted result set and one that hasn't been queried at all yet going solely by a null check of `nextRecord`. We could possibly nullify `resultSet` but I'm not sure that'd make things much more readable, and it may be more dangerous given how that field is used by superclasses.

Reply:
Well, a few things to discuss here.
- we can already use nullness of `resultSet` to distinguish whether the query was not yet executed (resultSet == null) vs end of resultset (resultSet.next is false or throws). I don't think there is need to explicitly null out the resultSet for this.
- the more serious issue I see here is that we are depending on the vendor-dependent behavior of `ResultSet.next()` returning false instead of throwing when hitting end of result set. The scenario I am thinking of that would fail is if the JDBC driver threw on `ResultSet.next()` at end of results, and the result set contained only one result. The next() implementation here would call `ResultSet.next()` twice, throwing on the second call. Since the offset determination is based on looking ahead by one, I don't think we can get around the need to advance the cursor in resultset ahead by at least one. That means we need to work around the possibility that vendors decide to throw on end of resultset.

Reply:
I believe we would have to explicitly nullify `resultSet`, since the only other time it's nullified is when resetting the querier, and during green-path execution that only occurs after `TableQuerier::next` returns `false`. If we reach the end of the result set but still have a cached record to return, it won't make sense to return `false`.

It doesn't seem advantageous to add in extra conditions for the nullability of the result set field that superclasses will have to deal with just to accommodate one special case in the `TimestampQuerier` class. Tracking logic specific to timestamp mode exclusively in the `TimestampQuerier` class seems preferable, even if it requires an additional field. This logic is also documented carefully (nearly every line has a comment) and tested thoroughly. Even if we can eliminate a field or remove a few lines of code, I don't believe it would be an improvement if the means to do so is to mutate inherited state that's utilized through three layers of abstraction.